packages/dds/tree/docs/wip/identifiers.md

```diff
@@ -0,0 +1,73 @@
+# SharedTree Node Identifiers
+
+## Why Identifiers
+
+Why use identifiers at all? SharedTree has paths, are not those already sufficient "references" or "handles" to nodes in the tree? No, a node's identifier is related to, but not the same thing as, the path to that node in the tree. A path points to a location within a tree, but it does not necessarily capture any semantic meaning about the identity of the node it points to. Consider a node A which is deleted, and then consider a node B which is inserted in the same place. Node A and Node B have the same path (although they are in different revisions) but they are not the same node; they should therefore not share the same identifier. Likewise, consider a node A which is moved from location X to location Y in the tree. Node A has a different path after the move than it did before, but semantically it is the "same node"; it should have the same identifier in both locations.
+
+## Proposal Overview
+
+Identifiers are stored under nodes in fields. A SharedTree has an optional index, the IdentifierIndex. The IdentifierIndex allows a user to provide an identifier and retrieve a cursor to a node with that identifier. An identifier is a special type (a numerically compressed node identifier). The index maintains a mapping from each identifier to a path. If history is enabled, then each identifier maps to a list of tuples of [path, revision]. The index is parameterized over a field key, e.g. "identifier", and will consider all nodes that have this field to be identified by the data in that field. That data ought to be an identifier; if it is not, the index can either ignore it or fail.
```

Comment:
About "If history is enabled, then each identifier maps to a list of tuples of [path, revision].", it's not clear to me whether that API or a `(identifier, revision) -> path` API would be more desirable. Potentially both?

Reply:
I think the design that Craig and I have been leaning towards is to have "Index Views", which are an index at a particular revision. In legacy SharedTree we had RevisionView which would be analogous to a ForestIndexView. Each index must be able to generate a view of itself for a given revision. This means it can be a copy-on-write structure under the hood, like RevisionView was. Or, it can have snapshot(s) of its state and apply deltas to move around to different revisions. Or, it can copy its entire state for every revision I suppose :) I'm not sure exactly what index APIs emerge from that but hopefully they end up being sane... what do you think of that approach?